core/src/main/scala/kafka/message/ByteBufferMessageSet.scala

```diff
@@ -139,11 +139,15 @@ object ByteBufferMessageSet {
         val newMessage = new Message(buffer, wrapperMessageTimestampOpt, wrapperMessageTimestampTypeOpt)
 
         // Inner message and wrapper message must have same magic value
-        if (newMessage.magic != wrapperMessage.magic)
-          throw new IllegalStateException(s"Compressed message has magic value ${wrapperMessage.magic} " +
-            s"but inner message has magic value ${newMessage.magic}")
+        if (newMessage.magic != wrapperMessage.magic) {
+          // The old LogCleaner had a bug which could cause compressed messages sets for magic=0
+          // to be placed in a wrapped message with magic=1.
+          if (!ignorePartiallyConverted || newMessage.magic != MagicValue_V0 || wrapperMessage.magic != MagicValue_V1)
```

Comment:
In theory, the bug can also generate a message with wrapper magic of 1 and inner magic of 0. For example: (1) an 0.10 message is published to a compacted topic; (2) message format is set to 0.9; (3) compaction happens. Since we are loosening the check, perhaps we can accommodate that case as well. Also, I am not sure how useful ignorePartiallyConverted is. It seems it's simpler to just always allow the reader to go through if the magic between the wrapper and inner don't match.